### ISSUE TITLE
s3 ext_pillar fails if key and keyid are not provided

### ISSUE THREAD

**Issue by rterbush** [saltstack/salt#29912]
Even though the doc suggest that an s3 ext_pillar will get credentials from the EC2 meta-data, this appears not to be working in 2015.8.3.

It is possible this may be a handoff issue with another part of the code as I see the following error in the logs (that is not really reported as an ERROR)
```
2015-12-21 22:44:37,691 [urllib3.connectionpool                              ][DEBUG   ][20158] "GET /?prefix=pillar HTTP/1.1" 403 None
```

@stanislavb may know something about this?

**Comment by jfindlay** [comment]
@rterbush, thanks for reporting.

**Comment by stanislavb** [comment]
Hi

Are you by any chance using salt on EC2 instances outside of us-east? I haven't dug deep in this issue, but the following config worked for me in us-east-1 region of AWS:

```
# Pillar backend
ext_pillar:
  - s3:
      bucket: salt-pillar-bucket
```

However outside of us-east-1 (in my case eu-west-1) I had to add an additional configuration line:

```
# Pillar backend
ext_pillar:
  - s3:
      bucket: salt-pillar-bucket
      service_url: s3-eu-west-1.amazonaws.com
```

It looks like requests to get data from S3 always use a correct region specific endpoint, but before that the request to create temporary credentials puts the "service_url" endpoint in header, resulting in S3 complaining about invalid signature and returning 403.

@rterbush If you can, please check if the workaround works.

**Comment by rterbush** [comment]
@stanislavb thanks for taking a look at this.

Unfortunately, the service_url parameter does not solve the problem for me. I am working in us-west-2 region. I've tried explicitly setting this parameter as well as letting the system figure it out. Neither work.

Let me know what else I can do to help figure this out.

The config I am currently trying is:
```
ext_pillar:
  - s3:
      bucket: salt-pillar-bucket
      service_url: s3-us-west-2.amazonaws.com
      prefix: pillar
```

**Comment by rterbush** [comment]
Any other thoughts here or things that I may or may not be doing that would cause this issue?

**Comment by stanislavb** [comment]
Since we're dealing with a Permission Denied from S3, as a sanity check I will mention that the IAM policy applied to the instance should have s3:GetObject permissions to objects in the bucket and s3:ListBucket permission to the bucket itself.

It would be awesome to have a more complete bit of log. Here is a part of my log with "log_level_logfile: debug" in master config. I have not tried using a prefix. Bucket name and some authentication headers are redacted below.

```
2016-01-01 01:12:07,926 [salt.utils.lazy                                     ][DEBUG   ][24092] LazyLoaded s3.ext_pillar
2016-01-01 01:12:07,926 [salt.loaded.int.pillar.s3                           ][DEBUG   ][24092] S3 bucket cache file /var/cache/salt/master/pillar_s3fs/salt-pillar-bucket--files.cache is expired, mtime_diff=-1346813.09294s, expiration=30s
2016-01-01 01:12:07,927 [salt.loaded.int.pillar.s3                           ][DEBUG   ][24092] Refreshing S3 buckets pillar cache file
2016-01-01 01:12:07,927 [salt.loaded.int.pillar.s3                           ][DEBUG   ][24092] Single environment per bucket mode
2016-01-01 01:12:07,928 [urllib3.connectionpool                              ][INFO    ][24092] Starting new HTTP connection (1): 169.254.169.254
2016-01-01 01:12:07,931 [urllib3.connectionpool                              ][DEBUG   ][24092] "GET /latest/meta-data/iam/security-credentials/ HTTP/1.1" 200 64
2016-01-01 01:12:07,932 [urllib3.connectionpool                              ][INFO    ][24092] Starting new HTTP connection (1): 169.254.169.254
2016-01-01 01:12:07,934 [urllib3.connectionpool                              ][DEBUG   ][24092] "GET /latest/meta-data/iam/security-credentials/StackName-InstanceRole-18GTYBHMOTX1I HTTP/1.1" 200 930
2016-01-01 01:12:07,935 [urllib3.connectionpool                              ][INFO    ][24092] Starting new HTTP connection (1): 169.254.169.254
2016-01-01 01:12:07,937 [urllib3.connectionpool                              ][DEBUG   ][24092] "GET /latest/dynamic/instance-identity/document HTTP/1.1" 200 419
2016-01-01 01:12:07,938 [salt.utils.s3                                       ][DEBUG   ][24092] S3 Request: https://salt-pillar-bucket.s3-eu-west-1.amazonaws.com/?prefix=
2016-01-01 01:12:07,938 [salt.utils.s3                                       ][DEBUG   ][24092] S3 Headers::
2016-01-01 01:12:07,938 [salt.utils.s3                                       ][DEBUG   ][24092]     Authorization: AWS4-HMAC-SHA256 Credential=<redacted>/20160101/eu-west-1/s3/aws4_request, SignedHeaders=host;x-amz-date;x-amz-security-token, Signature=<redacted>
2016-01-01 01:12:07,938 [urllib3.connectionpool                              ][INFO    ][24092] Starting new HTTPS connection (1): salt-pillar-bucket.s3-eu-west-1.amazonaws.com
2016-01-01 01:12:08,005 [urllib3.connectionpool                              ][DEBUG   ][24092] "GET /?prefix= HTTP/1.1" 200 None
2016-01-01 01:12:08,006 [salt.utils.s3                                       ][DEBUG   ][24092] S3 Response Status Code: 200
```

```
$ salt --versions-report
Salt Version:
           Salt: 2015.8.3

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
         Python: 2.6.9 (unknown, Dec  8 2015, 18:23:29)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist:
        machine: x86_64
        release: 4.1.7-15.23.amzn1.x86_64
```

@rterbush Do you think you can produce the information above so we have a more complete chain of events the pillar system is doing?

**Comment by rterbush** [comment]
@stanislavb 
As mentioned, I have access to this S3 resource via aws cli and I have confirmed that the role policy reflects the requirement for s3:Get* and s3:List*

Here is the log:
```
2016-01-02 00:05:57,420 [salt.utils.lazy                                     ][DEBUG   ][22763] LazyLoaded s3.ext_pillar
2016-01-02 00:05:57,421 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22763] S3 bucket cache file /var/cache/salt/master/pillar_s3fs/lhn-staging-role-manage-pillar-files.cache is expired, mtime_diff=-180015.006895s, expiration=30s
2016-01-02 00:05:57,421 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22763] Refreshing S3 buckets pillar cache file
2016-01-02 00:05:57,421 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22763] Single environment per bucket mode
2016-01-02 00:05:57,423 [urllib3.connectionpool                              ][INFO    ][22763] Starting new HTTP connection (1): 169.254.169.254
2016-01-02 00:05:57,426 [urllib3.connectionpool                              ][DEBUG   ][22763] "GET /latest/meta-data/iam/security-credentials/ HTTP/1.1" 200 24
2016-01-02 00:05:57,428 [urllib3.connectionpool                              ][INFO    ][22763] Starting new HTTP connection (1): 169.254.169.254
2016-01-02 00:05:57,432 [urllib3.connectionpool                              ][DEBUG   ][22763] "GET /latest/meta-data/iam/security-credentials/LHN-ManagementServerRole HTTP/1.1" 200 930
2016-01-02 00:05:57,433 [salt.utils.lazy                                     ][DEBUG   ][22816] LazyLoaded local_cache.prep_jid
2016-01-02 00:05:57,439 [urllib3.connectionpool                              ][INFO    ][22763] Starting new HTTP connection (1): 169.254.169.254
2016-01-02 00:05:57,442 [urllib3.connectionpool                              ][DEBUG   ][22763] "GET /latest/dynamic/instance-identity/document HTTP/1.1" 200 419
2016-01-02 00:05:57,444 [salt.utils.s3                                       ][DEBUG   ][22763] S3 Request: https://lhn-staging-role-manage.s3-us-west-2.amazonaws.com/?prefix=pillar
2016-01-02 00:05:57,444 [salt.utils.s3                                       ][DEBUG   ][22763] S3 Headers::
2016-01-02 00:05:57,444 [salt.utils.s3                                       ][DEBUG   ][22763]     Authorization: AWS4-HMAC-SHA256 Credential=/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-date;x-amz-security-token, Signature=
2016-01-02 00:05:57,447 [urllib3.connectionpool                              ][INFO    ][22763] Starting new HTTPS connection (1): lhn-staging-role-manage.s3-us-west-2.amazonaws.com
2016-01-02 00:05:57,479 [urllib3.connectionpool                              ][DEBUG   ][22804] "GET /?prefix=pillar HTTP/1.1" 403 None
2016-01-02 00:05:57,480 [salt.utils.s3                                       ][DEBUG   ][22804] S3 Response Status Code: 403
2016-01-02 00:05:57,483 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22804] Writing S3 buckets pillar cache file
2016-01-02 00:05:57,483 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22804] S3 bucket retrieved pillars {'base': {'lhn-staging-role-manage': []}}
2016-01-02 00:05:57,484 [salt.loaded.int.pillar.s3                           ][INFO    ][22804] Syncing local pillar cache from S3...
2016-01-02 00:05:57,484 [salt.loaded.int.pillar.s3                           ][INFO    ][22804] Sync local pillar cache from S3 completed.
2016-01-02 00:05:57,486 [urllib3.connectionpool                              ][DEBUG   ][22763] "GET /?prefix=pillar HTTP/1.1" 403 None
2016-01-02 00:05:57,486 [salt.utils.s3                                       ][DEBUG   ][22763] S3 Response Status Code: 403
2016-01-02 00:05:57,489 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22763] Writing S3 buckets pillar cache file
2016-01-02 00:05:57,490 [salt.loaded.int.pillar.s3                           ][DEBUG   ][22763] S3 bucket retrieved pillars {'base': {'lhn-staging-role-manage': []}}
2016-01-02 00:05:57,491 [salt.loaded.int.pillar.s3                           ][INFO    ][22763] Syncing local pillar cache from S3...
2016-01-02 00:05:57,491 [salt.loaded.int.pillar.s3                           ][INFO    ][22763] Sync local pillar cache from S3 completed.
```

salt version info:
```
[root@manage srv]# salt --versions
Salt Version:
           Salt: 2015.8.3

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.3.1
         Python: 2.6.9 (unknown, Dec 17 2015, 01:08:55)
           RAET: Not Installed
        Tornado: 4.1
            ZMQ: 3.2.5
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: 0.20.3
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist:
        machine: x86_64
        release: 4.1.7-15.23.amzn1.x86_64
```
I've tried this both on a system running Amazon Linux and another running Centos 7. Python versions 2.6 and 2.7 respectively.

I've also tried this with the prefix and without with top.sls existing in the root of both configurations.

Thanks for the input here.

**Comment by rterbush** [comment]
Another piece of info worth repeating is that if I provide credentials in the configuration, it works.

**Comment by rterbush** [comment]
@stanislavb I have this working now. And unfortunately, don't have a lot to offer as to why it now works.

I made some changes on the policy and role assigned to this EC2 instance and discovered that even with a wide open '*' permission assigned in the policy, that both the command line aws client as well as salt still did not have access to this bucket. Only after going into the bucket configuration and deleting the bucket permissions for the 'admin' role did it break the logjam here and allow me to list and access the bucket from the master instance using aws-cli.

After fixing those issues up, I was able to get the pillar access to work as expected and have confirmed that it works for SSE file properties as well. Now if we could just get someone to merge the KMS support, we would be set.

Thanks again for bearing with me here and taking time to help debug.
